File: apps/web/src/code/python/processes/background_processes.py
import time
from os import getenv

from e2b import Sandbox

E2B_API_KEY = getenv("E2B_API_KEY")


def print_stdout(output):
    print(output.line)


def main():
    sandbox = Sandbox(template="base", api_key=E2B_API_KEY)

    # Start a server process in the background
    # We are not using `background_server.wait()` - that would wait for the process to finish running
    background_server = sandbox.process.start(  # $HighlightLine
        "python3 -m http.server 8000",  # $HighlightLine
        on_stdout=print_stdout,  # $HighlightLine
    )  # $HighlightLine

    # Wait for the server to be accessible
    time.sleep(1)

    # Start another process that creates request to server
    server_request = sandbox.process.start("curl localhost:8000")

    # Wait for the server request to finish running
    request_output = server_request.wait()

    # Stop the background process (it would otherwise run indefinitely)
    background_server.kill()  # $HighlightLine

    # Access the server output after the server process is killed
    server_output = background_server.output  # $HighlightLine

    sandbox.close()


main()

File: apps/web/src/code/js/processes/background_processes.js
import { Sandbox } from 'e2b'

const sandbox = await Sandbox.create({
  template: 'base',
  apiKey: process.env.E2B_API_KEY,
})

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

// Start a server process in the background
// We are not calling `await backgroundServer` - that would wait for the process to finish running
const backgroundServer = await sandbox.process.start({
  // $HighlightLine
  cmd: 'python3 -m http.server 8000', // $HighlightLine
  onStdout: (output) => console.log(output), // $HighlightLine
}) // $HighlightLine

// Wait for the server to be accessible
await sleep(1000)

// Start another process that creates a request to server
const serverRequest = await sandbox.process.start({
  cmd: 'curl localhost:8000',
})

// Wait for the server request to finish running
const requestOutput = await serverRequest.wait()

// Stop the background process (it would otherwise run indefinitely)
await backgroundServer.kill() // $HighlightLine

// Access the server output after the server process is killed
const serverOutput = backgroundServer.output // $HighlightLine

await sandbox.close()
